src/KOKKOS/metatensor_system_kokkos.cpp

```diff
@@ -0,0 +1,412 @@
+        auto [samples_values_unique, samples_inverse, _] = torch::unique_dim(
+            samples_values, /*dim=*/0, /*sorted=*/true, /*return_inverse=*/true, /*return_counts=*/false
+        );
+
+        auto permutation = torch::arange(samples_inverse.size(0), samples_inverse.options());
+        samples_inverse = samples_inverse.flip({0});
+        permutation = permutation.flip({0});
+
+        auto sample_indices = torch::empty(samples_values_unique.size(0), samples_inverse.options());
+        sample_indices.scatter_(0, samples_inverse, permutation);
+
+        // wrap into metatensor data structures
+        auto samples = torch::make_intrusive<metatensor_torch::LabelsHolder>(
+            std::vector<std::string>{"first_atom", "second_atom", "cell_shift_a", "cell_shift_b", "cell_shift_c"},
+            samples_values_unique
+        );
+
+        auto neighbor_list = torch::make_intrusive<metatensor_torch::TensorBlockHolder>(
+            interatomic_vectors_filtered_full_or_half.index_select(0, sample_indices).unsqueeze(-1),
+            samples->to(device),
+            std::vector<metatensor_torch::TorchLabels>{
+                metatensor_torch::LabelsHolder::create({"xyz"}, {{0}, {1}, {2}})->to(device),
+            },
+            metatensor_torch::LabelsHolder::create({"distance"}, {{0}})->to(device)
+        );
+
+        metatensor_torch::register_autograd_neighbors(system, neighbor_list, options_.check_consistency);
+        system->add_neighbor_list(cache.options, neighbor_list);
+    }
+
+}
+
+
+template<class DeviceType>
+void MetatensorSystemAdaptorKokkos<DeviceType>::setup_neighbors_no_remap(metatensor_torch::System& system) {
+    throw std::runtime_error("The metatensor/kk requires remap_pairs to be true");
+}
+
+
+template<class DeviceType>
+metatensor_torch::System MetatensorSystemAdaptorKokkos<DeviceType>::system_from_lmp(
+    bool do_virial,
+    bool remap_pairs,
+    torch::ScalarType dtype,
+    torch::Device device
+) {
+    auto total_n_atoms = atomKK->nlocal + atomKK->nghost;
+
+    auto atom_types_lammps_kokkos = atomKK->k_type.view<DeviceType>();
+    auto mapping = options_.types_mapping_kokkos;
+    Kokkos::View<int32_t*, Kokkos::LayoutRight, DeviceType> atom_types_metatensor_kokkos("atom_types_metatensor", total_n_atoms);
+
+    Kokkos::parallel_for(
+        "MetatensorSystemAdaptorKokkos::system_from_lmp::atom_types_mapping",
+        Kokkos::RangePolicy(0, total_n_atoms),
+        KOKKOS_LAMBDA(int i)
+    {
+        atom_types_metatensor_kokkos(i) = mapping(atom_types_lammps_kokkos(i));
+    });
+
+    atomic_types_ = torch::from_blob(
+        atom_types_metatensor_kokkos.data(),
+        {total_n_atoms},
+        torch::TensorOptions().dtype(torch::kInt32).device(device)
+    ).clone();  // clone because the original memory belongs to Kokkos and will be deallocated
```

Comment:
instead of the clone, we could use the trick of moving the data inside a lambda, and registering the lambda as a custom Tensor destructor. If this is fast/small enough, then it might not be worth it

Reply:
This array should be slim enough that a copy is irrelevant, but perhaps it might be worth it somewhere else (positions and/or force arrays, which are the largets ones we manipulate between torch and Kokkos). In general, however, my profiling shows that this sort of copy is very far down the things we should try to optimize